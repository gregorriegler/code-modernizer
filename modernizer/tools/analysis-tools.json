[
  {
    "name": "entry-point-analysis",
    "description": "No description available",
    "arguments": [
      "project_path: string - Path to the project file",
      "file_name: string - Name of the file to analyze",
      "workspaceLoader: IWorkspaceLoader"
    ]
  }
]