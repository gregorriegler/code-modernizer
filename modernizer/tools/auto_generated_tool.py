from .base_auto_generated_tool import BaseAutoGeneratedTool

class AutoGeneratedTool(BaseAutoGeneratedTool):
    """
    A tool that automatically generates its interface from the corresponding C# refactoring tool
    using reflection to read the tool's arguments and description.
    """
    
    def __init__(self, tool_name, runcommand, tool_data):
        super().__init__(tool_name, runcommand, tool_data, "Refactoring")
        
    def _get_project_path(self):
        """Get the project path for refactoring tools"""
        return 'refactoring-tools/RoslynRefactoring/RoslynRefactoring.csproj'