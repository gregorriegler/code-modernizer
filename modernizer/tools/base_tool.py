import subprocess

class BaseTool:
    name = ''
    
    def __init__(self):
        self.runcommand = None
        
    def execute(self, *args, **kwargs):
        raise NotImplementedError("Subclasses must implement execute()")
    
    def get_description_from_csharp(self, tool_name):
        try:
            result = subprocess.run(
                ['dotnet', 'run', '--project', 'refactoring-tools/RoslynRefactoring/RoslynRefactoring.csproj', '--', '--describe', tool_name],
                capture_output=True,
                text=True,
                timeout=10
            )
            if result.returncode == 0:
                return result.stdout.strip()
            else:
                return f"Error getting description: {result.stderr}"
        except Exception as e:
            return f"Error getting description: {str(e)}"
