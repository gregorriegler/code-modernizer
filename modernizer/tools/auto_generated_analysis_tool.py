from .base_tool import BaseTool

class AutoGeneratedAnalysisTool(BaseTool):
    """
    A tool that automatically generates its interface from the corresponding C# analysis tool
    using reflection to read the tool's arguments and description.
    """
    
    def __init__(self, tool_name, runcommand, tool_data):
        super().__init__()
        self.name = tool_name
        self.runcommand = runcommand
        self._tool_data = tool_data
        self._info = self._convert_tool_data_to_info(tool_data)
        
    @property
    def info(self):
        """Get complete tool information from JSON data"""
        return self._info
    
    def _convert_tool_data_to_info(self, tool_data):
        """Convert JSON tool data to the expected info format"""
        args = tool_data.get('arguments', [])
        
        return {
            'name': tool_data.get('name', self.name),
            'description': tool_data.get('description', f'Analysis tool: {self.name}'),
            'arguments': args
        }
        
    @property
    def description(self):
        """Get the tool description from the C# analysis tool"""
        return self.info.get('description', f'Analysis tool: {self.name}')
    
    @property
    def arguments(self):
        """Get argument information"""
        info = self.info
        return info.get('arguments', [])
    
    def get_usage_info(self):
        """Get comprehensive usage information for this tool"""
        data = self._tool_data
        usage_info = f"""Tool: {data['name']}
Description: {data['description']}

Usage: /{data['name']} <arguments>

Arguments:"""
        
        for arg in data.get('arguments', []):
            usage_info += f"\n  - {arg}"
            
        return usage_info
    
    
    def execute(self, args):
        """Execute the analysis tool with the provided arguments"""
        arg_list = args.split() if isinstance(args, str) and args else []
        
        return self.runcommand('dotnet', [
            'run', '--project', 'refactoring-tools/RoslynAnalysis/RoslynAnalysis.csproj',
            '--', self.name
        ] + arg_list)