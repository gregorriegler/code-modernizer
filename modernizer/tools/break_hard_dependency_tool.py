from .base_tool import BaseTool

class BreakHardDependencyTool(BaseTool):
    name = 'break-hard-dependency'
    
    def __init__(self, runcommand):
        super().__init__()
        self.runcommand = runcommand
        
    def execute(self, args):
        arg_list = args.split()
        return self.runcommand('dotnet', ['run', '--project', 'refactoring-tools/RoslynRefactoring/RoslynRefactoring.csproj', '--', 'break-hard-dependency'] + arg_list)