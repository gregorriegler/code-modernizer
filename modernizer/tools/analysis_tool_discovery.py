import os
import subprocess
import json
from .auto_generated_analysis_tool import AutoGeneratedAnalysisTool

class AnalysisToolDiscovery:
    def __init__(self, runcommand):
        self.runcommand = runcommand
        self.cache_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'analysis-tools.json')
    
    def discover_analysis_tools(self):
        """Automatically discover all available C# analysis tools from JSON with caching"""
        tools_data = self._load_from_cache()
        if tools_data is None:
            tools_data = self._fetch_and_cache_tools()
        
        return [AutoGeneratedAnalysisTool(tool_data['name'], self.runcommand, tool_data)
                for tool_data in tools_data]
    
    def _load_from_cache(self):
        """Load tools data from cache file if it exists"""
        if os.path.exists(self.cache_file):
            with open(self.cache_file, 'r') as f:
                return json.load(f)
        return None
    
    def _fetch_and_cache_tools(self):
        """Fetch tools data from C# project and cache it"""
        tool_library_dir = os.path.dirname(os.path.abspath(__file__))
        project_path = os.path.join(tool_library_dir, '../../refactoring-tools/RoslynAnalysis/RoslynAnalysis.csproj')
        
        result = subprocess.run(
            ['dotnet', 'run', '--project', project_path, '--', '--list-tools'],
            capture_output=True,
            text=True,
            timeout=10
        )
        
        if result.returncode != 0:
            raise RuntimeError(f"Failed to discover analysis tools: {result.stderr}")
        
        tools_data = json.loads(result.stdout.strip())
        
        with open(self.cache_file, 'w') as f:
            json.dump(tools_data, f, indent=2)
        
        return tools_data