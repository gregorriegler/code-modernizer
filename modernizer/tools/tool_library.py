import re
import subprocess

from .ls_tool import LsTool
from .cat_tool import CatTool
from .test_tool import TestTool
from .coverage_tool import CoverageTool
from .mutation_tool import MutationTool
from .auto_generated_tool import AutoGeneratedTool
from .revert_tool import RevertTool
from .create_file_tool import CreateFileTool
from .edit_file_tool import EditFileTool
from .refactoring_tool_discovery import RefactoringToolDiscovery
from .analysis_tool_discovery import AnalysisToolDiscovery

class ToolLibrary:
    def __init__(self):
        static_tools = self._create_static_tools()
        dynamic_tools = self._discover_dynamic_tools()
        
        self.tools = static_tools + dynamic_tools
        self._build_tool_dict()
    
    def _create_static_tools(self):
        """Create the core static tools that are always available."""
        return [
            LsTool(self.runcommand),
            CatTool(self.runcommand),
            TestTool(self.runcommand),
            CoverageTool(self.runcommand),
            MutationTool(self.runcommand),
            RevertTool(self.runcommand),
            CreateFileTool(self.runcommand),
            EditFileTool(self.runcommand)
        ]
    
    def _discover_dynamic_tools(self):
        """Discover and create dynamic tools from external sources."""
        refactoring_discovery = RefactoringToolDiscovery(self.runcommand)
        refactoring_tools = refactoring_discovery.discover_refactoring_tools()
        
        analysis_discovery = AnalysisToolDiscovery(self.runcommand)
        analysis_tools = analysis_discovery.discover_analysis_tools()
        
        return refactoring_tools + analysis_tools
    
    def _build_tool_dict(self):
        self.tool_dict = {tool.name: tool for tool in self.tools}
    
    def get_tool_info(self, tool_name=None):
        """Get comprehensive information about tools"""
        if tool_name:
            tool = self.tool_dict.get(tool_name)
            if not tool:
                return f"Tool '{tool_name}' not found. Available tools: {', '.join(self.tool_dict.keys())}"
            
            if hasattr(tool, 'get_usage_info'):
                return tool.get_usage_info()
            else:
                return f"Tool: {tool.name}\nDescription: {getattr(tool, 'description', 'No description available')}"
        else:
            info_lines = ["Available Tools:"]
            for tool in self.tools:
                description = getattr(tool, 'description', 'No description available')
                info_lines.append(f"  {tool.name}: {description}")
            return "\n".join(info_lines)
    
    def get_refactoring_tools_info(self):
        """Get detailed information about all refactoring tools"""
        refactoring_tools = [tool for tool in self.tools if isinstance(tool, AutoGeneratedTool)]
        if not refactoring_tools:
            return "No refactoring tools found."
        
        info_lines = ["Refactoring Tools (auto-discovered from C#):"]
        for tool in refactoring_tools:
            info_lines.append(f"\n{tool.get_usage_info()}")
        
        return "\n".join(info_lines)
        
    def parse_and_execute(self, text):
        pattern = r'^/([\w-]+)(?:\s+(.*))?$'
        
        # Handle multi-line commands by treating the entire text as one command
        if '\n' in text:
            # For multi-line text, extract command from first line but use entire text for arguments
            first_line = text.splitlines()[0].strip()
            match = re.match(pattern, first_line)
            if match:
                cmd, _ = match.groups()
                tool = self.tool_dict.get(cmd)
                if tool:
                    # Extract arguments from the entire text, preserving newlines
                    # Find the position after "/{cmd} " in the text
                    cmd_prefix = f"/{cmd} "
                    if text.startswith(cmd_prefix):
                        full_args = text[len(cmd_prefix):]
                        result = tool.execute(full_args.strip() if full_args else None)
                        return text, result['output']
                    else:
                        result = tool.execute(None)
                        return text, result['output']
                else:
                    return text, f"Unknown command: {cmd}"
        
        # Single-line processing (original behavior)
        for line in text.splitlines():
            line = line.strip()
            match = re.match(pattern, line)
            if match:
                cmd, arg = match.groups()
                tool = self.tool_dict.get(cmd)
                if tool:
                    result = tool.execute(arg.strip() if arg else None)
                    return line, result['output']
                else:
                    return line, f"Unknown command: {cmd}"
        return text.splitlines()[0].strip(), ""

    def runcommand(self, cmd, args=None, cwd=None):
        try:
            if args:
                if isinstance(args, str):
                    args = [args]
                command_line = [cmd] + args
    
            timeout = 300 if (args and 'stryker' in args) else 30
            
            result = subprocess.run(
                command_line,
                capture_output=True,
                text=True,
                encoding='utf-8',
                errors='replace',
                timeout=timeout,
                cwd=cwd
            )
    
            output = result.stdout.rstrip('\n')
            
            if result.stderr:
                stderr = result.stderr
                if output:  # Only add newline if there's stdout content
                    output += f"\nSTDERR: {stderr}"
                else:
                    output = f"STDERR: {stderr}"
    
            return {
                'success': result.returncode == 0,
                'output': output,
                'returncode': result.returncode
            }
        except subprocess.TimeoutExpired:
            return {'success': False, 'output': 'Command timed out (30s limit)', 'returncode': -1}
        except Exception as e:
            return {'success': False, 'output': f'Error: {str(e)}', 'returncode': -1}
