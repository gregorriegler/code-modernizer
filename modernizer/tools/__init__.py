from .base_tool import BaseTool
from .tool_library import ToolLibrary
from .cat_tool import CatTool
from .ls_tool import LsTool
from .revert_tool import RevertTool
from .test_tool import TestTool

__all__ = [
    'BaseTool',
    'ToolLibrary',
    'CatTool',
    'LsTool',
    'RevertTool',
    'TestTool'
]