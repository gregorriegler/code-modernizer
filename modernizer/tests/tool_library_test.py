import os
import sys

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from modernizer.tools.tool_library import ToolLibrary
from modernizer.tools.auto_generated_tool import AutoGeneratedTool
from .test_helpers import (
    create_temp_file,
    create_temp_directory_structure,
    verifyTool
)

library = ToolLibrary()

def test_ls_tool_basic_directory(tmp_path):
    directory_path, _, _, _, _ = create_temp_directory_structure(tmp_path)
    
    verifyTool(library, f"/ls {directory_path}")

def test_ls_tool_nonexistent_directory():
    verifyTool(library, "/ls /nonexistent/path")

def test_cat_tool_single_file(tmp_path):
    temp_file = create_temp_file(tmp_path, "test.txt", "Hello world\nSecond line\nThird line")
    
    verifyTool(library, f"/cat {temp_file}")

def test_cat_tool_nonexistent_file():
    verifyTool(library, "/cat /nonexistent/file.txt")

def test_cat_tool_empty_file(tmp_path):
    temp_file = create_temp_file(tmp_path, "empty.txt", "")
    
    verifyTool(library, f"/cat {temp_file}")

def test_unknown_command():
    verifyTool(library, "/unknown-command arg1 arg2")

def test_no_command_in_text():
    verifyTool(library, "This is just regular text without commands")

def test_tool_with_special_characters(tmp_path):
    special_file = create_temp_file(tmp_path, "special-file_name.txt", "Content with special chars: !@#$%^&*()")
    
    verifyTool(library, f"/cat {special_file}")

def test_auto_generated_tools_discovered():
    auto_generated_tools = [tool for tool in library.tools if isinstance(tool, AutoGeneratedTool)]
    discovered_tool_names = [tool.name for tool in auto_generated_tools]
    
    assert len(auto_generated_tools) > 1, f"Expected more than 1 auto-generated tool, found {len(auto_generated_tools)}"
    assert 'extract-method' in discovered_tool_names, f"extract-method not found in discovered tools: {discovered_tool_names}"

def test_create_tool_single_character_name(tmp_path):
    # Change to temp directory for test
    original_cwd = os.getcwd()
    os.chdir(tmp_path)
    
    try:
        verifyTool(library, "/create a")
        
        # Verify file was created
        assert os.path.exists("a"), "File 'a' should have been created"
        assert os.path.getsize("a") == 0, "File 'a' should be empty"
    finally:
        os.chdir(original_cwd)

def test_create_tool_simple_name_with_extension(tmp_path):
    # Change to temp directory for test
    original_cwd = os.getcwd()
    os.chdir(tmp_path)
    
    try:
        verifyTool(library, "/create test.txt")
        
        # Verify file was created
        assert os.path.exists("test.txt"), "File 'test.txt' should have been created"
        assert os.path.getsize("test.txt") == 0, "File 'test.txt' should be empty"
    finally:
        os.chdir(original_cwd)

def test_create_tool_single_character_content(tmp_path):
    # Change to temp directory for test
    original_cwd = os.getcwd()
    os.chdir(tmp_path)
    
    try:
        verifyTool(library, "/create test.txt a")
        
        # Verify file was created with correct content
        assert os.path.exists("test.txt"), "File 'test.txt' should have been created"
        with open("test.txt", "r") as f:
            content = f.read()
        assert content == "a", f"File 'test.txt' should contain 'a', but contains '{content}'"
    finally:
        os.chdir(original_cwd)