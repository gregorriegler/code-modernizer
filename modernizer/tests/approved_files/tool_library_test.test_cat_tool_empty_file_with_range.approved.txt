Command:
/cat /tmp/test_path/empty.txt 1-5

Result:
STDERR: cat: 'C:\Users\riegl\AppData\Local\Temp\pytest-of-riegl\pytest-496\test_cat_tool_empty_file_with_0\empty.txt 1-5': No such file or directory
